### PR TITLE
Command handling entity

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
Going to fix #319 in this PR.

File: server/src/main/java/org/spine3/server/command/HandlerRegistry.java [hunk 1]
```diff
@@ -102,14 +105,21 @@ private void checkDuplicates(MethodMap<CommandHandlerMethod> handlers) {
...
+                final CommandHandler handler = getHandler(commandClass);
+                final HandlerMethod alreadyRegistered = getHandlerMethod(handler.getClass(), commandClass.value());
                 throw new CommandHandlerAlreadyRegisteredException(commandClass,
                         alreadyRegistered.getFullName(),
                         entry.getValue().getFullName());
             }
         }
     }
 
+    private static CommandHandlerMethod getHandlerMethod(Class<? extends CommandHandler> clazz,
```
Comment (dmdashenkov) [inline server/src/main/java/org/spine3/server/command/HandlerRegistry.java]: Please, have more descriptive name for the `clazz` parameter.

File: server/src/main/java/org/spine3/server/aggregate/EventApplierMethod.java [hunk 2]
```diff
@@ -49,12 +52,22 @@
     }
 
     /**
+     * Returns the set of the event classes generated by the passed {@code Aggregate} class.
+     *
+     * @param clazz the class of the aggregate
+     * @return immutable set of event classes
+     */
+    static ImmutableSet<Class<? extends Message>> getEventClasses(Class<? extends Aggregate> clazz) {
```
Comment (dmdashenkov) [inline server/src/main/java/org/spine3/server/aggregate/EventApplierMethod.java]: I'm not sure that this is `EventApplierMethod` responsibility. Maybe we should extract this into a utility class or leave in the `Aggregate` class.

File: server/src/main/java/org/spine3/server/command/CommandHandler.java [hunk 3]
```diff
@@ -20,107 +20,66 @@
...
- * The first parameter is a command message. The second parameter is {@link CommandContext}.
- *
- * <p>The method returns an event message of the specific type, or {@code List} of messages
- * if it produces more than one event.
- *
- * <p>The method may throw one or more throwables derived from {@link FailureThrowable}.
- * Throwing a {@code FailureThrowable} indicates that the passed command cannot be handled
- * because of a business failure, which can be obtained via {@link FailureThrowable#getFailure()}.
+ * <p>A command handler does not have own state. So the state of the business
+ * model it changes is external to it. Even though such a behaviour may be needed in
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/command/CommandHandler.java]: There can be cases when it's really needed. We recommend another approach. That should be enough.

File: server/src/main/java/org/spine3/server/command/CommandBus.java [hunk 4]
```diff
@@ -218,22 +261,18 @@ public void post(Command command, StreamObserver<Response> responseObserver) {
...
-        if (dispatcherRegistry.hasDispatcherFor(commandClass)) {
-            dispatch(command);
-        } else if (handlerRegistry.handlerRegistered(commandClass)) {
-            invokeHandler(message, commandContext);
-        }
+    @VisibleForTesting
+    boolean isSupportedCommand(CommandClass commandClass) {
+        final boolean dispatcherRegistered = dispatcherRegistry.hasDispatcherFor(commandClass);
+        final boolean handlerRegistered = handlerRegistry.handlerRegistered(commandClass);
+        final boolean isSupported = dispatcherRegistered || handlerRegistered;
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/command/CommandBus.java]: Agreed. I've filed the issue #320. Let's improve it in another PR.

File: server/src/main/java/org/spine3/server/aggregate/EventApplierMethod.java (hunk 2 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/aggregate/EventApplierMethod.java]: It's dead code. Now used only in tests. Will remove.

**Pull request merged**